compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToUKernels.cpp

```diff
@@ -0,0 +1,263 @@
+  Value n0 = getDimAsI32(rewriter, loc, rhs, 2);
+  Value k0 = getDimAsI32(rewriter, loc, rhs, 3);
+  Value flagsVal = rewriter.create<arith::ConstantOp>(
+      loc, rewriter.getI32IntegerAttr(flags));
+  auto fn = getFnNameAndDefAttrs(ukernelName, rewriter, targetAttr);
+
+  // Create UKernel for AMD-AIE.
+  auto genericMicroKernelOp = rewriter.create<IREE::Codegen::UKernelGenericOp>(
+      loc, outType, fn.name, ValueRange{lhs, rhs}, out,
+      ValueRange{m, n, k, m0, n0, k0, flagsVal},
+      /*fn_def_attrs=*/rewriter.getDictionaryAttr(fn.defAttrs),
+      /*strided_outer_dims=*/rewriter.getIndexAttr(1));
+  return cast<IREE::Codegen::UKernelOpInterface>(
+      genericMicroKernelOp.getOperation());
+  return success();
+}
+
+using TargetPredicate = std::function<bool(IREE::HAL::ExecutableTargetAttr)>;
+
+template <typename OpType>
+struct LowerToUKernelPattern : OpRewritePattern<OpType> {
+  LowerToUKernelPattern(MLIRContext *context, TargetPredicate targetPredicate)
+      : OpRewritePattern<OpType>(context), targetPredicate(targetPredicate) {}
+
+  LogicalResult matchAndRewrite(OpType op,
+                                PatternRewriter &rewriter) const override {
+    if (targetPredicate &&
+        !targetPredicate(IREE::HAL::ExecutableTargetAttr::lookup(op))) {
+      return failure();
+    }
+    FailureOr<IREE::Codegen::UKernelOpInterface> ukernelOp =
+        matchDAGForUKernel(rewriter, op);
+    if (failed(ukernelOp)) {
+      return rewriter.notifyMatchFailure(
+          op, "failed to find microkernel op to replace with");
+    }
+    rewriter.replaceOp(op, ukernelOp.value()->getResults());
+    return success();
+  }
+
+  TargetPredicate targetPredicate;
+};
+
+}  // namespace
+
+void AMDAIELowerToUKernelsPass::runOnOperation() {
+  MLIRContext *context = &getContext();
+  RewritePatternSet patterns(context);
+  // Enabling a lowering of an op to a microkernel is a trade-off between the
+  // potential performance advantage of a microkernel over pure code generation
+  // for that op, and the potential benefits of fusions. Indeed, once an op
+  // lowered into a microkernel, it will never be fused at any MLIR level.
+  // Since microkernels are linked as bitcode, they will still undergo LTO-like
+  // optimization in their calling contexts, but we shouldn't expect this to
+  // achieve similar results as fusing structured ops.
+
+  // These patterns are unconditionally enabled, because we have strong evidence
+  // that it is difficult for codegen to consistently approach microkernels
+  // performance, and that consideration overrides the benefit of fusions for
+  // these ops.
+  auto allTargets = [](auto target) { return true; };
+  // TODO(avarma): There is no need to have microkernels for anything but the
+  // target-optimal tiled layout. So, while building the microkernel pipeline, I
+  // need to have a linalg.matmul -> linalg.mmt4d.
+  patterns.insert<LowerToUKernelPattern<linalg::Mmt4DOp>>(context, allTargets);
```

Comment:
I started off with `linalg.matmul`, but reverted to `linalg.mmt4d` based on my understanding that -> it makes sense to have microkernel only for tiled performant ISA specific implementations.

@MaheshRavishankar please let me know if we want to target `linalg.matmul` itself in this pass - which means we would be invoking this way earlier in the pipeline before the "tile+fuse" enchilada (since `linalg.matmul` gets lowered to a tiled `linalg.generic` in the pipeline).

Based on the discussion, I can reconfigure it to target `linalg.matmul` OR follow along with the `linalg.matmul -> linalg.mmt4d` comment I added as a `TODO`.

Reply:
Just a heads up that it might be a vector.contract in the near future

Reply:
I've had a discussion with @MaheshRavishankar and have addressed the PR accordingly.

The decision was to target a `linalg.generic` since that's what we get after the last `pack-and-transpose` pass via simple-pack pipeline.

Therefore, I've added a utility that matches a tiled `linalg.generic` op which implements a matmul and have added TODOs/comments to explain the rationale in the pass too.

One of the cases that I bumped into while implementing was the presence of more than one `linalg.generic`s in the dispatch - I've added an implementation that deals with it, have demonstrated it via a lit test too and have added a TODO that highlights a possible "better" way to deal with the same.

NOTE: This pass seems like would go right after the last `pack-and-transpose + canonicalize + cse`. This is because `memref` related replacements are too ugly to deal with and the pass would crash too since a bufferized DPS linalg doesn't really have a "result". By "result", I mean : `%result = linalg.generic ....`. And due to this the `replaceOp` would cry.

Reply:
We will have to balance the use of microkernels and generic vectorization path, maybe add a flag, or prefer microkernels over direct vectorization. `vector.contract` is a harder op to lower to microkernel cause you dont have a pointer value. So we will have to figure out an ABI for vector types.